Support running on macOS

btriller opened the pull request:
##### DOOL VERSION
`next`

##### SUMMARY
<!--- Describe the change here, including rationale and design decisions -->
Allow running `dool` on FreeBSD and macOS to allow easier development of plugins on these platform.

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

Before this change, `dool` just hangs.

scottchiefbaker commented:
Rather than making separate `ticks*` functions for each OS, wouldn't it be simpler to just make the existing `ticks()` function detect the OS and return appropriately? 

It looks like `ticks()` only gets called once on each invocation so performance isn't a huge deal.

scottchiefbaker commented:
Maybe something like this? I don't have MacOS to test though.

This prevents us from having to make changes to any of the rest of the Dool code right?

```python
def ticks():
    "Return the number of 'ticks' since bootup"

    uname = os.uname()

    # FreeBSD and MacOS
    if (uname.sysname in ['Darwin', 'FreeBSD']):
        import subprocess
        boottime = subprocess.check_output(['sysctl', '-n', 'kern.boottime'])
        boottime = boottime.decode().split()[3:7]
        boottime = float(f'{boottime[0].rstrip(",")}.{boottime[3]}')

        return time.time() - boottime

    # Linux
    try:
        for line in open('/proc/uptime', 'r').readlines():
            l = line.split()
            if len(l) < 2: continue
            return float(l[0])
    except:
        for line in dopen('/proc/stat').readlines():
            l = line.split()
            if len(l) < 2: continue
            if l[0] == 'btime':
                return time.time() - int(l[1])
```

scottchiefbaker commented:
Looking into this a little more, I'm not 100% what `ticks()` does. It says it returns ticks since bootup, but that's not true. Instead it returns a float of SECONDS since bootup. The doc string should probably be updated to reflect that.

This is **only** used to calculate the `step` value which is essentially which iteration were one since invocation.

btriller committed:
Use human-readable attribute names

Named attributes are available since Python 3.3 [1]

[1] https://docs.python.org/3/library/os.html#os.uname

btriller committed:
Support running on macOS

btriller commented:
> Rather than making separate `ticks*` functions for each OS, wouldn't it be simpler to just make the existing `ticks()` function detect the OS and return appropriately?
> 
> It looks like `ticks()` only gets called once on each invocation so performance isn't a huge deal.

I somehow thought, this would be executed in every iteration. 😞

scottchiefbaker commented:
I've been working on `dool` for six plus years now and I still learn stuff about it. The guts are pretty insane, so it takes some deep dives to really figure things out.

I had no idea a feature like this was even possible. This means that FreeBSD and MacOS have the same (or very similar) `/proc/` file structure as Linux?

btriller commented:
> I had no idea a feature like this was even possible. This means that FreeBSD and MacOS have the same (or very similar) `/proc/` file structure as Linux?

Unfortunately they don't. That's why the fiddling with `sysctl` for `boottime`. So most plugins are useless on these platforms.
